### ISSUE TITLE
Allow configuration of private keys to use your own for TLS management

### ISSUE THREAD

**Issue by francislavoie** [caddyserver/caddy#2854]
## 1. What would you like to have changed?

I think it would be useful for some users to allow specifying the private key to use during certificate management instead of Caddy always generating its own keys.

## 2. Why is this feature a useful, necessary, and/or important addition to this project?

In certain situations, people may want to configure clients with certificate/pubkey pinning. Currently this is a bit tricky without this feature.

Certificate pinning is obviously not ideal because of the fast renewal time, and pinning against the intermediate (Let's Encrypt) doesn't really offer much benefit. Pinning against the public key is a potentially more viable option because Caddy keeps using the same private key indefinitely.

The issue is that Caddy's key management is currently pretty opaque, it's not intended to have users mess with the storage. If someone is changing servers or is coming from a different server (nginx, etc.) and want to do pubkey pinning, then they don't really have any good options when moving to Caddy.

Letting the user specify the private key does entail more risks because the user may have inadvertently leaked the key when moving it between servers, but it would allow more flexibility.

## 3. What alternatives are there, or what are you doing in the meantime to work around the lack of this feature?

Right now, I guess the user could write the private key manually to Caddy's storage using the same filename as Caddy would expect, but that's undocumented and not recommended. The other option is when they transition, they could have the client pin against Caddy's new private key and the old key, then later remove pinning of the old key, but that's also not ideal.

## 4. Please link to any relevant issues, pull requests, or other discussions.

N/A, some private discussion in Slack

**Comment by sarge** [comment]
In V2 you can specify the tls certificate using this format.

```caddyfile
test.site.com: 443 {
 tls /somepath/site.cer /somepath/site.key {
  }
}
```

Was this what you were looking for?

**Comment by francislavoie** [comment]
No, if you read the request, this is asking for the key to be user-defined, but not the certificate - the certificate should still be managed by Let's Encrypt. 

The syntax you gave is for specifying your own cert/key pair that you've had signed externally from Caddy. 

This issue is asking for an option to tell Caddy which private key to use instead of letting it generate its own. This makes it more reliable to use the public portion of that key for pinning in a client (e.g. mobile app, such that they only handshake with the server that has that private key)

**Comment by francislavoie** [comment]
I think this could be implemented by setting `certmagic.Config.KeySource` to a struct that reads a key from file in the `Generate()` function instead of randomly generating one, if configured.

**Comment by pauloappbr** [comment]
Hey everyone,

I’ve completed an implementation for this feature and opened the corresponding pull requests.

To keep the architecture modular and aligned with the existing design, the solution is split into two parts:

- caddyserver/certmagic#371  
  Introduces `FileKeyGenerator`, enabling CertMagic to load and use a pre-existing PEM-encoded private key from the filesystem.

- caddyserver/caddy#7509
  Adds the `key_file` directive (global and per-site) and integrates it into the TLS automation policies.

**Dependency note:**  
The Caddy PR depends on the CertMagic changes being merged first. Once `FileKeyGenerator` is available in CertMagic, Caddy can leverage it to allow users to supply their own private keys for automated certificate management (ACME or Internal).

This addresses the long-standing request for key continuity across deployments and infrastructure migrations, while preserving Caddy’s existing automated TLS behavior by default.

I’d appreciate any feedback or suggestions.
